File: src/fastapi_420/storage/lua/token_bucket.lua
-- Token bucket rate limiting with Redis
-- KEYS[1] = rate limit key (hash)
-- ARGV[1] = limit
-- ARGV[2] = window seconds
-- ARGV[3] = cost
-- ARGV[4] = current timestamp
-- ARGV[5] = burst capacity (optional, defaults to limit)

local key = KEYS[1]
local limit = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local cost = tonumber(ARGV[3])
local now = tonumber(ARGV[4])
local capacity = tonumber(ARGV[5]) or limit

local rate = limit / window  -- tokens per second

local bucket = redis.call('HMGET', key, 'tokens', 'last_update')
local tokens = tonumber(bucket[1])
local last_update = tonumber(bucket[2])

if not tokens then
    tokens = capacity
    last_update = now
end

-- Refill tokens
local elapsed = now - last_update
tokens = math.min(capacity, tokens + elapsed * rate)

if tokens < cost then
    local needed = cost - tokens
    local retry_after = math.ceil(needed / rate)
    local reset_time = now + retry_after
    return {0, limit, 0, reset_time, retry_after}
end

tokens = tokens - cost
redis.call('HMSET', key, 'tokens', tokens, 'last_update', now)
redis.call('EXPIRE', key, math.ceil(capacity / rate) + window + 1)

local remaining = math.floor(tokens)
local reset_time = now + (capacity - tokens) / rate

return {1, limit, remaining, reset_time, 0}

File: src/fastapi_420/algorithms/token_bucket.py
"""Token bucket rate limiting algorithm."""

import time
from threading import Lock
from typing import Dict

from fastapi_420.algorithms.base import BaseAlgorithm, RateLimitResult


class TokenBucketAlgorithm(BaseAlgorithm):
    """Token bucket rate limiting.
    
    Allows burst traffic up to bucket size, then enforces steady rate.
    Good for allowing occasional bursts while maintaining average rate.
    """
    
    def __init__(self, limit: int, window_seconds: int, burst: Optional[int] = None):
        super().__init__(limit, window_seconds)
        # Tokens per second
        self.rate = limit / window_seconds
        # Bucket size (allow bursts)
        self.capacity = burst or limit
        self._buckets: Dict[str, tuple[float, float]] = {}  # key -> (tokens, last_update)
        self._lock = Lock()
    
    def _get_bucket(self, key: str) -> tuple[float, float]:
        """Get or create bucket for key."""
        with self._lock:
            if key not in self._buckets:
                self._buckets[key] = (float(self.capacity), time.time())
            return self._buckets[key]
    
    def _refill(self, tokens: float, last_update: float) -> float:
        """Refill tokens based on elapsed time."""
        now = time.time()
        elapsed = now - last_update
        new_tokens = tokens + elapsed * self.rate
        return min(new_tokens, self.capacity)
    
    def check_limit(self, key: str, cost: int = 1) -> RateLimitResult:
        with self._lock:
            if key not in self._buckets:
                self._buckets[key] = (float(self.capacity), time.time())
            
            tokens, last_update = self._buckets[key]
            tokens = self._refill(tokens, last_update)
            
            if tokens < cost:
                # Calculate retry time
                needed = cost - tokens
                retry_after = int(needed / self.rate) + 1
                reset_time = time.time() + retry_after
                return RateLimitResult(
                    allowed=False,
                    limit=self.limit,
                    remaining=0,
                    reset_time=reset_time,
                    retry_after=retry_after
                )
            
            tokens -= cost
            self._buckets[key] = (tokens, time.time())
            remaining = int(tokens)
            reset_time = time.time() + (self.capacity - tokens) / self.rate
            
            return RateLimitResult(
                allowed=True,
                limit=self.limit,
                remaining=remaining,
                reset_time=reset_time
            )
    
    def reset(self, key: str) -> None:
        with self._lock:
            if key in self._buckets:
                del self._buckets[key]
    
    def get_current_usage(self, key: str) -> int:
        with self._lock:
            if key not in self._buckets:
                return 0
            tokens, last_update = self._buckets[key]
            tokens = self._refill(tokens, last_update)
            return int(self.capacity - tokens)